src/Language/Pirouette/Example/IsUnity.hs

```diff
@@ -36,78 +36,89 @@ checkOk =
+    (\(x : a) (xs : List a) .
+      match_List @a y0 @Bool False (\(y : a) (ys : List a) . and (eq x y) (listEq @a eq xs ys)))
+
+contains : forall a . (a -> Bool) -> List a -> Bool
+contains @a eq x0 = 
+  match_List @a x0 @Bool
+    False
+    (\(x : a) (xs : List a) . or (eq x) (contains @a eq xs))
+
+data Pair (x : *) (y : *)
   = P : x -> y -> Pair x y
 
-fun pairEq : all (a : Type)(b : Type) . (a -> a -> Bool) -> (b -> b -> Bool) -> Pair a b -> Pair a b -> Bool
-  = /\(a : Type)(b : Type) . \(eqA : a -> a -> Bool) (eqB : b -> b -> Bool) (x : Pair a b) (y : Pair a b)
-  . match_Pair @a @b x @Bool
-      (\(x0 : a) (x1 : b) . match_Pair @a @b y @Bool
-        (\(y0 : a) (y1 : b) . and (eqA x0 y0) (eqB x1 y1)))
-
-data Maybe (x : Type)
+pairEq :
+  forall a . forall b .
+  (a -> a -> Bool) ->
+  (b -> b -> Bool) ->
+  Pair a b -> Pair a b -> Bool
+pairEq @a @b eqA eqB x y =
+  match_Pair @a @b x @Bool
+    (\(x0 : a) (x1 : b) . match_Pair @a @b y @Bool
+      (\(y0 : a) (y1 : b) . and (eqA x0 y0) (eqB x1 y1)))
+
+data Maybe (x : *)
   = Just : x -> Maybe x
   | Nothing : Maybe x
 
-fun maybeSum : all (x : Type) . Maybe x -> Maybe x -> Maybe x
-  = /\ (x : Type) . \(mx : Maybe x)(my : Maybe x)
-  . match_Maybe @x mx @(Maybe x)
-      (\(jx : x) . Just @x jx)
-      my
+maybeSum : forall x . Maybe x -> Maybe x -> Maybe x
+maybeSum @x mx my =
+  match_Maybe @x mx @(Maybe x)
+    (\(jx : x) . Just @x jx)
+    my
 
-fun isJust : all (x : Type) . Maybe x -> Bool
-  = /\ (x : Type) . \(mx : Maybe x)
-  . match_Maybe @x mx @Bool (\(jx : x) . True) False
+isJust : forall x . Maybe x -> Bool
+isJust @x mx =
+  match_Maybe @x mx @Bool (\(jx : x) . True) False
 
-data KVMap (k : Type) (v : Type)
+data KVMap (k : *) (v : *)
   = KV : List (Pair k v) -> KVMap k v
 
-fun toList : all (k : Type)(v : Type) . KVMap k v -> List (Pair k v)
-  = /\(k : Type)(v : Type) . \(m : KVMap k v) . match_KVMap @k @v m @(List (Pair k v)) (\(x : List (Pair k v)) . x)
+toList : forall k . forall v . KVMap k v -> List (Pair k v)
+toList @k @v m = match_KVMap @k @v m @(List (Pair k v)) (\(x : List (Pair k v)) . x)
 
-fun lkupOne : all (k : Type)(v : Type) . (k -> Bool) -> Pair k v -> Maybe v
-  = /\(k : Type)(v : Type) . \(predi : k -> Bool)(m : Pair k v)
-  . match_Pair @k @v m @(Maybe v)
-      (\(fst : k) (snd : v) . if @(Maybe v) predi fst then Just @v snd else Nothing @v)
+lkupOne : forall k . forall v . (k -> Bool) -> Pair k v -> Maybe v
+lkupOne @k @v predi m =
+  match_Pair @k @v m @(Maybe v)
+    (\(fst : k) (snd : v) . if @(Maybe v) predi fst then Just @v snd else Nothing @v)
 
-fun lkup : all (k : Type)(v : Type) . (k -> k -> Bool) -> KVMap k v -> k -> Maybe v
-  = /\(k : Type)(v : Type) . \(eq : k -> k -> Bool)(m : KVMap k v) (tgt : k)
-  . match_KVMap @k @v m @(Maybe v)
-      (foldr @(Pair k v) @(Maybe v) (\(pk : Pair k v) . maybeSum @v (lkupOne @k @v (eq tgt) pk)) (Nothing @v))
+lkup : forall k . forall v . (k -> k -> Bool) -> KVMap k v -> k -> Maybe v
```

Comment:
Can't we write `forall k v . (k -> k -> Bool) ... `?

Reply:
We can. Some of these had evaded my refactoring pass. I also changed some `data Something (a: *) (b : *)` into `data Something a b` that were still left in their desugared form. This is now done.